changeloggenutils/merged_release.go

```diff
@@ -0,0 +1,252 @@
+    if earlierVersion == nil {
+        return []Version{*laterVersion}
+    }
+    var i, j = -1, -1
+    // otherRepoReleasesSorted is sorted from highest semver to lowest semver v1.7.0 -> v1.0.0
+    for idx, release := range otherRepoReleasesSorted {
+        if release == *laterVersion {
+            i = idx
+        }
+        if release == *earlierVersion {
+            j = idx
+            break
+        }
+        // Don't look for dependent versions across major / minor versions
+        if i != -1 && (release.Major != earlierVersion.Major || release.Minor != earlierVersion.Minor) {
+            j = idx
+            break
+        }
+    }
+    if i < 0 || j < 0 || j+1 < i {
+        return nil
+    }
+    return otherRepoReleasesSorted[i:j]
+}
+
+func getGithubReleaseMarkdownLink(tag, repoOwner, repo string) string {
+    return fmt.Sprintf("[%s](https://github.com/%s/%s/releases/tag/%s)", tag, repoOwner, repo, tag)
+}
+
+// Looks for
+// repoOwner/otherRepo v1.x.x-x
+// in go.mod string (goMod) and returns the package version
+func getPkgVersionFromGoMod(goMod, repoOwner, repo string) (*Version, error){
+    semverRegex := fmt.Sprintf("%s/%s\\s+(v((([0-9]+)\\.([0-9]+)\\.([0-9]+)(?:-([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?)(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?))",
+        repoOwner, repo)
+    regex := regexp.MustCompile(semverRegex)
+    // Find version of open source dependency
+    matches := regex.FindStringSubmatch(goMod)
+    if len(matches) < 2 {
+        return nil, fmt.Errorf("unable to find dependency")
+    }
+    return ParseVersion(matches[1])
+}
+
+/*
+ Enterprise changelogs have the option to "merge" open source changelogs. The following function retrieves and returns
+ the open source version that an enterprise version depends on. It checks out the enterprise go.mod at the release version
+ and looks for the open source dependency ({RepoOwner}/{DependentRepo}).
+*/
+func (g *MergedReleaseGenerator) GetOpenSourceDependency(enterpriseVersion *Version) (*Version, error) {
+    files, err := githubutils.GetFilesFromGit(context.TODO(), g.client, g.opts.RepoOwner, g.opts.MainRepo, enterpriseVersion.String(), "go.mod")
+    if err != nil {
+        return nil, fmt.Errorf("error fetching dependency for enterprise version %s: %s", enterpriseVersion.String(), err.Error())
+    }
+    if len(files) < 1 {
+        return nil, fmt.Errorf("unable to find go.mod file in enteprise repository")
+    }
+    content, err := files[0].GetContent()
+    if err != nil {
+        return nil, err
+    }
+
+    return getPkgVersionFromGoMod(content, g.opts.RepoOwner, g.opts.DependentRepo)
+}
+
+// HOF that returns a Dependency function. The returned function generates
+// changelogs faster than the default but requires passing in the github token.
+func GetOSDependencyFunc(repoOwner, enterpriseRepo, osRepo, githubToken string) (DependencyFn, error){
```

Comment:
How much faster is this vs the alternate depedencyFn?

Reply:
It's a matter of how many versions have been released. This clones an entire repo in memory, which takes about 10 seconds, then the look up of each go.mod file for ~200 versions takes less than 1 second. 
The default method queries github API on every request, so for ~100 versions it takes 10 seconds, but the query time per version is much higher than the in-memory func.